chain/network/src/test_loop.rs

```diff
@@ -1,4 +1,301 @@
+/// reverse order.
+///
+/// Examples of custom handlers
+/// - Override handler to skip sending messages to or from a specific client.
+/// - Override handler to simulate more network delays.
+/// - Override handler to modify data and simulate malicious behavior.
+#[derive(Default)]
+pub struct TestLoopPeerManagerActor {
+    handlers: Vec<NetworkRequestHandler>,
+}
+
+impl Actor for TestLoopPeerManagerActor {}
+
+impl TestLoopPeerManagerActor {
+    /// Create a new TestLoopPeerManagerActor with default handlers for client, partial_witness, and shards_manager.
+    /// Note that we should be able to access the senders for these actors from the data type.
+    pub fn new<'a, T>(clock: Clock, account_id: &AccountId, datas: &'a Vec<T>) -> Self
+    where
+        AccountId: From<&'a T>,
+        ClientSenderForTestLoopNetwork: From<&'a T>,
+        PartialWitnessSenderForNetwork: From<&'a T>,
+        Sender<ShardsManagerRequestFromNetwork>: From<&'a T>,
+    {
+        let mut actor = Self { handlers: Vec::new() };
+        vec![
+            network_message_to_client_handler(&account_id, make_sender_map(datas)),
+            network_message_to_partial_witness_handler(&account_id, make_sender_map(datas)),
+            network_message_to_shards_manager_handler(clock, &account_id, make_sender_map(datas)),
+            network_message_to_state_snapshot_handler(),
+        ]
+        .into_iter()
+        .for_each(|handler| actor.register_override_handler(handler));
+        actor
+    }
+
+    /// Register a new handler to override the default handlers.
+    pub fn register_override_handler(&mut self, handler: NetworkRequestHandler) {
+        self.handlers.push(handler);
+    }
+}
+
+// Helper function to create a map of senders from a list of data.
+// Converts Vec<Data> to HashMap<AccountId, Sender>
+fn make_sender_map<'a, T, U>(datas: &'a Vec<T>) -> HashMap<AccountId, U>
+where
+    AccountId: From<&'a T>,
+    U: From<&'a T>,
+{
+    let mut senders = HashMap::new();
+    for data in datas.iter() {
+        senders.insert(data.into(), data.into());
+    }
+    senders
+}
+
+impl Handler<SetChainInfo> for TestLoopPeerManagerActor {
+    fn handle(&mut self, _msg: SetChainInfo) {}
+}
+
+impl Handler<PeerManagerMessageRequest> for TestLoopPeerManagerActor {
+    fn handle(&mut self, msg: PeerManagerMessageRequest) -> PeerManagerMessageResponse {
+        let PeerManagerMessageRequest::NetworkRequests(request) = msg else {
+            panic!("Unexpected message: {:?}", msg);
+        };
+
+        // Iterate over the handlers in reverse order to allow for overriding the default handlers.
+        let mut request = Some(request);
+        for handler in self.handlers.iter().rev() {
+            if let Some(new_request) = handler(request.take().unwrap()) {
+                request = Some(new_request);
+            } else {
+                return PeerManagerMessageResponse::NetworkResponses(NetworkResponses::NoResponse);
```

Comment:
Side note: I always thought that NoResponse means that the message was unhandled and something went wrong, but it meant the opposite all the time.
We could add comment what `NetworkResponses::NoResponse` actually means in some other PR.

Reply:
Added a quick comment here atleast to signal successful handling.